netem/src/lib.rs

```diff
@@ -0,0 +1,535 @@
+//! Sans-IO network emulator inspired by Linux netem.
+//!
+//! This crate provides a network emulator that can simulate:
+//! - Latency and jitter
+//! - Packet loss (random or bursty via Gilbert-Elliot model)
+//! - Packet duplication
+//! - Packet reordering
+//! - Rate limiting
+//!
+//! # Sans-IO Pattern
+//!
+//! This implementation follows the Sans-IO pattern: packets go in with timestamps,
+//! and decisions come out (drop, delay, duplicate). The caller handles actual I/O
+//! and timing.
+//!
+//! # Example
+//!
+//! ```
+//! use std::time::{Duration, Instant};
+//! use str0m_netem::{Netem, NetemConfig, Input, Output, LossModel, RandomLoss, Probability};
+//!
+//! let config = NetemConfig::new()
+//!     .latency(Duration::from_millis(50))
+//!     .jitter(Duration::from_millis(10))
+//!     .loss(LossModel::Random(RandomLoss::new(Probability::new(0.01))))
+//!     .seed(42);
+//!
+//! let mut netem: Netem<Vec<u8>> = Netem::new(config);
+//!
+//! // Send a packet
+//! let now = Instant::now();
+//! netem.handle_input(Input::Packet(now, vec![1, 2, 3]));
+//!
+//! // Poll for output
+//! while let Some(output) = netem.poll_output() {
+//!     match output {
+//!         Output::Timeout(when) => {
+//!             // Wait until `when` and call handle_input with Input::Timeout
+//!         }
+//!         Output::Packet(data) => {
+//!             // Send the packet
+//!         }
+//!     }
+//! }
+//! ```
+
+mod config;
+mod loss;
+
+pub use config::{GilbertElliot, LossModel, NetemConfig, Probability, RandomLoss};
+
+use std::cmp::{Ordering, Reverse};
+use std::collections::BinaryHeap;
+use std::time::{Duration, Instant};
+
+use fastrand::Rng;
+
+use loss::LossState;
+
+/// Trait for getting the length of packet data (used for rate limiting).
+pub trait WithLen {
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+impl<T: AsRef<[u8]>> WithLen for T {
+    fn len(&self) -> usize {
+        self.as_ref().len()
+    }
+}
+
+/// Input events to the network emulator.
+#[derive(Debug)]
+pub enum Input<T> {
+    /// A timeout has occurred at the given instant.
+    Timeout(Instant),
+
+    /// A packet arrived at the given instant with the given data.
+    Packet(Instant, T),
+}
+
+/// Output events from the network emulator.
+#[derive(Debug)]
+pub enum Output<T> {
+    /// Request a timeout at the given instant.
+    Timeout(Instant),
+
+    /// A packet is ready to be sent.
+    Packet(T),
+}
+
+/// A queued packet waiting to be sent.
+#[derive(Debug)]
+struct QueuedPacket<T> {
+    /// When this packet should be sent.
+    send_at: Instant,
+
+    /// The packet data.
+    data: T,
+
+    /// Ever increasing counter to break ties when send_at is the same.
+    packet_index: u64,
+}
+
+/// Sans-IO network emulator.
+pub struct Netem<T> {
```

Comment:
Should this struct be first in the file?

Reply:
Yes